### PR TITLE
[Debugger] Optimize project API usages

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/30247 - Revisit project API usages and introduce project cache.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugSourceType.java [hunk 1]
```diff
@@ -21,5 +21,6 @@
  */
 public enum DebugSourceType {
     PACKAGE,
-    SINGLE_FILE
+    SINGLE_FILE,
+    DEPENDENCY,
```
Comment (suganyasuven) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugSourceType.java:25]:
```suggestion
    DEPENDENCY
```

Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugSourceType.java]: Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/30249/commits/e777c1ca7a62b119b0fcc5d6b6f2b4ae7fa1005f.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java [hunk 2]
```diff
@@ -31,22 +31,26 @@
...
  */
 public class SuspendedContext {
 
+    private final ExecutionContext executionContext;
     private final VirtualMachineProxyImpl attachedVm;
     private final ThreadReferenceProxyImpl owningThread;
     private final StackFrameProxyImpl frame;
-    private final Project project;
-    private final DebugSourceType sourceType;
+    private Project project;
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java:52]: Shall we make `project` to `final` variable
```suggestion
    private final Project project;
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java [hunk 3]
```diff
@@ -55,16 +59,14 @@
     private ClassLoaderReference classLoader;
     private DebugExpressionCompiler debugCompiler;
 
-    SuspendedContext(Project project, VirtualMachineProxyImpl vm, ThreadReferenceProxyImpl threadRef,
+    SuspendedContext(ExecutionContext executionContext, ThreadReferenceProxyImpl threadRef,
                      StackFrameProxyImpl frame) {
-        this.attachedVm = vm;
+        this.executionContext = executionContext;
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java:64]: Seems like there is no usage for `executionContext ` here. Shall we remove it?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugProjectCache.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        this.loadedProjects = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Returns the project instance which contains the given file path, from the project cache.
+     *
+     * @param filePath source root of the Ballerina project that need to be retrieved.
+     * @return project instance.
+     */
+    public Project getProjectFor(Path filePath) {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugProjectCache.java:49]: Shall we change the method name to `getProject`?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java [hunk 5]
```diff
@@ -31,22 +31,26 @@
...
 import static org.ballerinalang.debugadapter.DebugSourceType.SINGLE_FILE;
+import static org.ballerinalang.debugadapter.utils.PackageUtils.computeProjectKindAndRoot;
 import static org.ballerinalang.debugadapter.utils.PackageUtils.getFileNameFrom;
 
 /**
  * Context holder for debug suspended state related information.
  */
 public class SuspendedContext {
 
+    private final ExecutionContext executionContext;
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java:48]: Refer comment [here](https://github.com/ballerina-platform/ballerina-lang/pull/30249/files#r622845304)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/30249/commits/b757b9606646c2a17a343cbd0fc80347236265a6

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/30249/commits/b757b9606646c2a17a343cbd0fc80347236265a6

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/SuspendedContext.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/30249/commits/b757b9606646c2a17a343cbd0fc80347236265a6

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugProjectCache.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugProjectCache.java]: Changed in https://github.com/ballerina-platform/ballerina-lang/pull/30249/commits/f48d00ae36895cef8da5a5e7f05957056dbd3a93

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**